sdk/typescript/src/providers/json-rpc-provider.ts

```diff
@@ -54,7 +99,110 @@ export class JsonRpcProvider extends Provider {
     public skipDataValidation: boolean = false
   ) {
     super();
+
     this.client = new JsonRpcClient(endpoint);
+
+    this.wsEndpoint = getWebsocketUrl(endpoint);
+
+    const socketOptions = { reconnect_interval: 3000, autoconnect: false };
+    this.wsClient = new WsRpcClient(this.wsEndpoint, socketOptions);
+    this.setupSocket();
+  }
+
+  private setupSocket() {
+    this.wsClient.on('open', () => {
+      if(this.wsTimeout != 0) {
+        clearTimeout(this.wsTimeout);
+        this.wsTimeout = 0;
+      }
+      this.wsConnectionState = ConnectionState.Connected;
+      // underlying websocket is private, but we need it
+      // to access messages sent by the node
+      (this.wsClient as any).socket
+        .on('message', this.onSocketMessage.bind(this));
+    });
+
+    this.wsClient.on('close', () => {
+      this.wsConnectionState = ConnectionState.NotConnected;
+    });
+
+    this.wsClient.on('error', console.error);
+  }
+
+  private async connect(): Promise<void> {
+    if (this.wsConnectionState === ConnectionState.Connected)
```

Comment:
What if the client is actively connecting and this method gets called?

Reply:
its a good question, we could hang onto the connection promise and return that again?